File: apps/docs/src/prework/examples/rule-data.ts
import type { Reason, Request } from "./procedural";

type RuleData = Readonly<{
  field: keyof Request;
  expected: boolean;
  reason: Reason;
}>;

// この例では、規則を関数ではなくデータで表す。
export const rules: readonly RuleData[] = [
  { field: "equipmentChecked", expected: true, reason: "EquipmentNotChecked" },
  { field: "flareAlert", expected: false, reason: "FlareAlertActive" },
];

export function check(request: Request): readonly Reason[] {
  return rules
    .filter((rule) => request[rule.field] !== rule.expected)
    .map((rule) => rule.reason);
}

File: apps/docs/src/prework/examples/rust/rule_data.rs
use crate::procedural::{Reason, Request};

#[derive(Clone, Copy)]
pub enum Field {
    EquipmentChecked,
    FlareAlert,
}

pub struct RuleData {
    pub field: Field,
    pub expected: bool,
    pub reason: Reason,
}

pub const RULES: [RuleData; 2] = [
    RuleData {
        field: Field::EquipmentChecked,
        expected: true,
        reason: Reason::EquipmentNotChecked,
    },
    RuleData {
        field: Field::FlareAlert,
        expected: false,
        reason: Reason::FlareAlertActive,
    },
];

pub fn read_field(request: &Request, field: Field) -> bool {
    match field {
        Field::EquipmentChecked => request.equipment_checked,
        Field::FlareAlert => request.flare_alert,
    }
}

pub fn check(request: &Request) -> Vec<Reason> {
    RULES
        .iter()
        .filter(|rule| read_field(request, rule.field) != rule.expected)
        .map(|rule| rule.reason)
        .collect()
}
